crates/circuit/src/register.rs

```diff
@@ -0,0 +1,1054 @@
+                    crate::slice::SequenceIndex::Int(_) => {
+                        slf.getitem_inner(key)?.next().map(PyBit).into_py_any(py)
+                    }
+                    _ => Ok(PyList::new(py, slf.getitem_inner(key)?.map(PyBit))?
+                        .into_any()
+                        .unbind()),
+                }
+            }
+            _ => Ok(PyList::new(py, slf.getitem_inner(key)?.map(PyBit))?
+                .into_any()
+                .unbind()),
+        }
+    }
+
+    /// Find the index of the provided bit within this register.
+    fn index(slf: Bound<'_, Self>, bit: Bound<PyBit>) -> PyResult<usize> {
+        let err = || -> PyResult<usize> {
+            Err(PyValueError::new_err(format!(
+                "Bit {} not found in Register {}.",
+                bit,
+                slf.as_super().repr()?,
+            )))
+        };
+        let slf_borrowed = slf.borrow();
+        let bit_borrowed = bit.borrow();
+        let bit_inner = &bit_borrowed.0;
+        match bit_inner {
+            BitInfo::Owned { index, .. } => {
+                if slf_borrowed.0.contains(bit_inner) {
+                    Ok((*index) as usize)
+                } else {
+                    err()
+                }
+            }
+            BitInfo::Anonymous { .. } => match slf_borrowed.0.as_ref() {
+                RegisterInfo::Owning(..) => err(),
+                RegisterInfo::Alias { bits, .. } => {
+                    bits.get_index_of(bit_inner)
+                        .ok_or(PyValueError::new_err(format!(
+                            "Bit {} not found in Register {}.",
+                            bit,
+                            slf.repr()?,
+                        )))
+                }
+            },
+        }
+    }
+
+    /// Make object hashable, based on the name and size to hash.
+    fn __hash__(slf: Bound<'_, Self>) -> PyResult<isize> {
+        let borrowed = slf.borrow();
+        (slf.get_type(), borrowed.name(), borrowed.size())
+            .into_bound_py_any(slf.py())?
+            .hash()
+    }
+
+    /// Two Registers are the same if they are of the same type
+    /// (i.e. quantum/classical), and have the same name and size. Additionally,
+    /// if either Register contains new-style bits, the bits in both registers
+    /// will be checked for pairwise equality. If two registers are equal,
+    /// they will have behave identically when specified as circuit args.
+    ///
+    /// Args:
+    ///     other (Register): other Register
+    ///
+    /// Returns:
+    ///     bool: `self` and `other` are equal.
+    fn __eq__(slf: Bound<'_, Self>, other: Bound<'_, Self>) -> PyResult<bool> {
+        if slf.is(&other) {
+            return Ok(true);
+        }
+        let slf_borrow = slf.borrow();
+        let other_borrow = other.borrow();
+        Ok(slf.get_type().eq(other.get_type())?
+            && slf.repr()?.to_string().eq(&other.repr()?.to_string())
```

Comment:
I'm not sure we should be using a string comparison for this since we have borrowed versions of both object.

Reply:
Fixed in 883d82334.